File: app/src/main/java/com/example/ugcssample/drone/DroneBridgeBase.java
package com.example.ugcssample.drone;

import android.content.Context;
import android.hardware.usb.UsbAccessory;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.example.ugcssample.utils.AppScheduledExecutorService;
import com.example.ugcssample.utils.ThreadUtils;

public abstract class DroneBridgeBase implements DroneBridge {

    protected static final AppScheduledExecutorService WORKER = ThreadUtils.newSingleThreadScheduledExecutor(DroneBridge.class);

    protected final Context context;
    protected final LocalBroadcastManager lbm;
    protected final UsbAccessoryManager usbAccessoryManager;

    DroneBridgeBase(Context context) {
        this.context = context;
        this.lbm = LocalBroadcastManager.getInstance(context);
        UsbAccessoryManager.Listener usbAccessoryManagerListener = new UsbAccessoryManager.Listener() {

            @Override
            public void onUsbAccessoryConnected(final UsbAccessory usbAccessory) {
                WORKER.submit(() -> DroneBridgeBase.this.onUsbAccessoryConnectedSync(usbAccessory));
            }

            @Override
            public void onUsbAccessoryDisconnected() {
                WORKER.submit(DroneBridgeBase.this::onUsbAccessoryDisconnectedSync);
            }
        };
        this.usbAccessoryManager = new UsbAccessoryManagerImplV2(context, WORKER, usbAccessoryManagerListener);
    }

    abstract void onUsbAccessoryConnectedSync(UsbAccessory usbAccessory);

    abstract void onUsbAccessoryDisconnectedSync();

    protected Context getContext() {
        return context;
    }
}

File: app/src/main/java/com/example/ugcssample/drone/DroneBridgeBase.kt
package com.example.ugcssample.drone

import android.content.Context
import android.hardware.usb.UsbAccessory
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.example.ugcssample.utils.AppScheduledExecutorService
import com.example.ugcssample.utils.ThreadUtils

abstract class DroneBridgeBase internal constructor(protected val context: Context) : DroneBridge {
    protected val lbm: LocalBroadcastManager = LocalBroadcastManager.getInstance(context)
    protected val usbAccessoryManager: UsbAccessoryManager
    abstract fun onUsbAccessoryConnectedSync(usbAccessory: UsbAccessory?)
    abstract fun onUsbAccessoryDisconnectedSync()

    companion object {
        val WORKER: AppScheduledExecutorService = ThreadUtils.newSingleThreadScheduledExecutor(
            DroneBridge::class.java
                                                                                )
    }

    init {
        val usbAccessoryManagerListener: UsbAccessoryManager.Listener =
            object : UsbAccessoryManager.Listener {
                override fun onUsbAccessoryConnected(usbAccessory: UsbAccessory?) {
                    WORKER.submit { onUsbAccessoryConnectedSync(usbAccessory) }
                }

                override fun onUsbAccessoryDisconnected() {
                    WORKER.submit { onUsbAccessoryDisconnectedSync() }
                }
            }
        usbAccessoryManager =
            UsbAccessoryManagerImplV2(context, WORKER, usbAccessoryManagerListener)
    }
}